Fix operations on SparseArrays of Integer types other than Int

rohitvarkey opened the pull request:
On current master, operations on `SparseArrays` of anything other than the system `Int` gives the following error

```julia
julia> using SparseArrays

julia> S = sparse(UInt8[1, 2, 3], UInt8[1, 2, 3], UInt8[1, 2, 3])
3×3 SparseMatrixCSC{UInt8,UInt8} with 3 stored entries:
  [1, 1]  =  0x01
  [2, 2]  =  0x02
  [3, 3]  =  0x03

julia> sum(S, 2)
ERROR: MethodError: no method matching _mapreducezeros(::typeof(identity), ::typeof(Base.add_sum), ::Type{UInt8}, ::UInt8, ::UInt64)
Closest candidates are:
  _mapreducezeros(::Any, ::Any, ::Type{T}, ::Int64, ::Any) where T at /Users/rohitvarkey/Code/julia/usr/share/julia/site/v0.7/SparseArrays/src/sparsematrix.jl:1590
  _mapreducezeros(::Any, ::typeof(+), ::Type{T}, ::Int64, ::Any) where T at /Users/rohitvarkey/Code/julia/usr/share/julia/site/v0.7/SparseArrays/src/sparsematrix.jl:1623
  _mapreducezeros(::Any, ::typeof(*), ::Type{T}, ::Int64, ::Any) where T at /Users/rohitvarkey/Code/julia/usr/share/julia/site/v0.7/SparseArrays/src/sparsematrix.jl:1625
Stacktrace:
 [1] _mapreducecols!(::typeof(identity), ::typeof(Base.add_sum), ::Array{UInt64,2}, ::SparseMatrixCSC{UInt8,UInt8}) at /Users/rohitvarkey/Code/julia/usr/share/julia/site/v0.7/SparseArrays/src/sparsematrix.jl:1670
 [2] _mapreducedim!(::typeof(identity), ::typeof(Base.add_sum), ::Array{UInt64,2}, ::SparseMatrixCSC{UInt8,UInt8}) at /Users/rohitvarkey/Code/julia/usr/share/julia/site/v0.7/SparseArrays/src/sparsematrix.jl:1687
 [3] mapreducedim! at ./reducedim.jl:240 [inlined]
 [4] mapreducedim at ./reducedim.jl:272 [inlined]
 [5] sum at ./reducedim.jl:612 [inlined]
 [6] sum(::SparseMatrixCSC{UInt8,UInt8}, ::Int64) at ./reducedim.jl:614
 [7] top-level scope

julia> versioninfo()
Julia Version 0.7.0-DEV.3743
Commit 8da832fb69* (2018-02-07 22:19 UTC)
Platform Info:
  OS: macOS (x86_64-apple-darwin16.4.0)
  CPU: Intel(R) Core(TM) i5-5257U CPU @ 2.70GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-3.9.1 (ORCJIT, broadwell)
Environment:
```

This change fixes this error. `julia` built with commit gives me the following.

```julia
julia> using SparseArrays

julia> S = sparse(UInt8[1, 2, 3], UInt8[1, 2, 3], UInt8[1, 2, 3])
3×3 SparseMatrixCSC{UInt8,UInt8} with 3 stored entries:
  [1, 1]  =  0x01
  [2, 2]  =  0x02
  [3, 3]  =  0x03

julia> sum(S, 2)
3×1 Array{UInt64,2}:
 0x0000000000000001
 0x0000000000000002
 0x0000000000000003

julia> versioninfo()
Julia Version 0.7.0-DEV.3744
Commit 8a4f0c6e53* (2018-02-08 04:25 UTC)
Platform Info:
  OS: macOS (x86_64-apple-darwin16.4.0)
  CPU: Intel(R) Core(TM) i5-5257U CPU @ 2.70GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-3.9.1 (ORCJIT, broadwell)
Environment:
```

andreasnoack commented on stdlib/SparseArrays/src/sparsematrix.jl:
```diff
@@ -1586,7 +1586,7 @@ Base.reducedim_initarray0(A::SparseMatrixCSC, region, v0, ::Type{R}) where {R} =
 
 # General mapreduce
-function _mapreducezeros(f, op, ::Type{T}, nzeros::Int, v0) where T
+function _mapreducezeros(f, op, ::Type{T}, nzeros::U, v0) where {T, U<:Integer}
```
Maybe a better solution is to convert `nzeros` to `Int` before calling these methods in order to avoid compiling a specialized version of this method.

rohitvarkey replied:
Thanks for the feedback @andreasnoack! I thought of that approach, but what would happen in that case for systems with a smaller `Int` than the type of `nzeros`?

eg: `Int` is `Int32` and `nzeros` has a value of `Int64` that overflows conversion to an `Int32`

andreasnoack replied:
I'm pretty sure that this code is not even remotely robust in that case. E.g. https://github.com/JuliaLang/julia/blob/834cc44d9eb519f374dec1f38ca104e55f6bc9ca/stdlib/SparseArrays/src/sparsematrix.jl#L1610 since
```julia
julia> length(speye(50000))
-1794967296
```
Getting this to work reliably on 32 bit systems will require a bit more work. Hence I still think it is better to avoid the unnecessary method compilation by converting to `Int` before calling this function.

Sacha0 commented on stdlib/SparseArrays/src/sparsematrix.jl:
```diff
@@ -1586,7 +1586,7 @@ Base.reducedim_initarray0(A::SparseMatrixCSC, region, v0, ::Type{R}) where {R} =
 
 # General mapreduce
-function _mapreducezeros(f, op, ::Type{T}, nzeros::Int, v0) where T
+function _mapreducezeros(f, op, ::Type{T}, nzeros::U, v0) where {T, U<:Integer}
```
Perhaps `nzeros::U, v0) where {T, U<:Integer}` -> `nzeros::Integer, v0) where T`?

rohitvarkey replied:
I was under the impression that approach would not generate efficient code because of the abstract type but I just checked the docs and found that my understanding was wrong! Thanks @Sacha0! I'll update the PR with this.

fredrikekre commented:
I think we can simply go with:
```diff
diff --git a/stdlib/SparseArrays/src/sparsematrix.jl b/stdlib/SparseArrays/src/sparsematrix.jl
index f710fa9..1688715 100644
--- a/stdlib/SparseArrays/src/sparsematrix.jl
+++ b/stdlib/SparseArrays/src/sparsematrix.jl
@@ -1667,7 +1667,7 @@ function _mapreducecols!(f, op, R::AbstractArray, A::SparseMatrixCSC{Tv,Ti}) whe
         end
     end
     @inbounds for i = 1:m
-        R[i, 1] = _mapreducezeros(f, op, Tv, rownz[i], R[i, 1])
+        R[i, 1] = _mapreducezeros(f, op, Tv, Int(rownz[i]), R[i, 1])
     end
     R
 end
```

It would also be great if you could add the testcase from OP in the bottom of https://github.com/JuliaLang/julia/blob/master/stdlib/SparseArrays/test/sparse.jl with a reference to this PR.

rohitvarkey committed:
convert nzeros to Int before calling _mapreducezeros

rohitvarkey commented:
I've updated the PR with the suggestion from @andreasnoack and @fredrikekre to convert `nzeros` to `Int` before calling `mapreducezeros` and added the testcase at the bottom of sparse.jl

Sacha0 reviewed (approved):
Much thanks for the great first contribution @rohitvarkey! Also, a warm welcome to the JuliaLang/julia repo! :)

Sacha0 commented:
(Absent objections or requests for time, I plan to merge these changes tomorrow morning PST. Best!)

rohitvarkey commented:
@Sacha0 @andreasnoack @fredrikekre Thanks for all the help on this!
